src/osd/OSD.cc

```diff
@@ -10029,17 +10060,35 @@ bool OSD::maybe_override_options_for_qos(const std::set<std::string> *changed)
         }
       }
     } else { // if (changed != nullptr) (osd boot-up)
-      // Override the default recovery max active and max backfills to
-      // higher values based on the type of backing device (hdd/ssd).
-      // This section is executed only during osd boot-up.
+      // Override the default recovery max active (hdd & ssd) and max
+      // backfills to mClock defaults. This section is executed only
+      // during osd boot-up.
       for (auto opt : recovery_qos_defaults) {
-        cct->_conf.set_val_default(opt.first, std::to_string(opt.second));
```

Comment:
Note that this is not addressed in https://github.com/athanatos/ceph/tree/sjust/wip-mclock-49975 as I'm not sure I'm following what's going on.

Reply:
This fix (commit 4126d02ce68d5edda595d0a97e83f8fdcfb2ed7b) is needed to deal with two default values for 
'osd_max_backfills'. The built-in default is set to 1 and that is used if 'wpq' scheduler is enabled. The other default is 
for mClock and is set to 3. The mClock defaults are maintained within `recovery_qos_defaults` map and 
is used to overwrite the built-in default of 1 using set_val_default().

The opt.second as part of set_max is to retain the mClock defaults in case of osd restarts *and* if there was no user 
initiated change to this config before the restart. But If the user had previously changed the value before the
 osd restart (using config set ...), then this fix also ensures that the new value is retained.

But all this seemingly weird logic should go away once we make mClock the default *and* remove/disable 'wpq' in
the code.

Reply:
That doesn't really answer my question.  Why wouldn't my proposed change above work?

Reply:
I am not sure if you looking at the latest version of the code (Commit 06be4504a5b5f6f9baa622b949cf77139f810775). If you recall, the recovery and
backfills limit modification for mClock is gated by another option called `osd_mclock_override_recovery_settings`. So a user would have to set this override
first before attempting to modify any recovery/backfill limits.

You version of the code is in the osd restart scenario. In your version, you get the current value of 'osd_max_backfill' using get_val(). This will either return the built-in default or the user initiated change. Assuming there was no change made, get_val() would return the built-in default which is 1 and later  set_max to 1 which is not what we want. We want the mClock default of 3 to be set and that's why `opt.second` needs to be used. AFAIK set_val_default() doesn't persist the value on the mon store and is lost on osd restart.

If the user had set the backfills limit using "config set", then that value is retained even across osd restarts. The logic in this section of the code detects that the
current value is different from the built-in default and mClock default (opt.second) and avoids making any change. So set_val_default() is not called in this scenario.

Reply:
May be the standalone test TEST_backfill_limit_adjustment_mclock() as part of the same commit will help in understanding the scenario.

Reply:
> You version of the code is in the osd restart scenario. In your version, you get the current value of 'osd_max_backfill' using get_val(). This will either return the built-in default or the user initiated change. Assuming there was no change made, get_val() would return the built-in default which is 1 and later set_max to 1 which is not what we want. We want the mClock default of 3 to be set and that's why `opt.second` needs to be used. AFAIK set_val_default() doesn't persist the value on the mon store and is lost on osd restart.



Originally, the code was:

```
      for (auto opt : recovery_qos_defaults) {
        cct->_conf.set_val_default(opt.first, std::to_string(opt.second));
        if (opt.first == "osd_max_backfills") {
          service.local_reserver.set_max(opt.second);
          service.remote_reserver.set_max(opt.second);
```

I am suggesting that this becomes:

```
      for (auto opt : recovery_qos_defaults) {
        cct->_conf.set_val_default(opt.first, std::to_string(opt.second));
        if (opt.first == "osd_max_backfills") {
          auto conf_max_backfills = cct->_conf.get_val<uint64_t>("osd_max_backfills");
          service.local_reserver.set_max(conf_max_backfills);
          service.remote_reserver.set_max(conf_max_backfills);
```

Because we called set_val_default, get_val should return the overridden value 3, not 1.